src/stream/src/executor/mview/materialize.rs

```diff
@@ -274,11 +443,82 @@ impl<S: StateStore, SD: ValueRowSerde> MaterializeExecutor<S, SD> {
                         | ConflictBehavior::DoUpdateIfNotNull => {
                             self.state_table.write_chunk(chunk.clone());
                             self.state_table.try_flush().await?;
+
+                            // For refreshable materialized views, also write to staging table during refresh
+                            if let Some(ref mut refresh_args) = self.refresh_args
+                                && refresh_args.is_refreshing
+                            {
+                                let key_chunk =
+                                    chunk.clone().project(self.state_table.pk_indices());
+                                tracing::trace!(
+                                    staging_chunk = %key_chunk.to_pretty(),
+                                    input_chunk = %chunk.to_pretty(),
+                                    "writing to staging table"
+                                );
+                                if cfg!(debug_assertions) {
+                                    // refreshable source should be append-only
+                                    assert!(key_chunk.ops().iter().all(|op| op == &Op::Insert));
+                                }
+                                refresh_args.staging_table.write_chunk(key_chunk.clone());
+                                refresh_args.staging_table.try_flush().await?;
+                            }
+
                             Message::Chunk(chunk)
                         } // ConflictBehavior::DoUpdateIfNotNull => unimplemented!(),
                     }
                 }
                 Message::Barrier(b) => {
+                    // Handle refresh mutations for refreshable materialized views
+                    if let Some(ref mut refresh_args) = self.refresh_args {
+                        if let Some(m) = b.mutation.as_deref() {
+                            tracing::debug!(?m, "barrier mutation received");
+                        }
+                        match b.mutation.as_deref() {
+                            Some(Mutation::RefreshStart {
+                                table_id: refresh_table_id,
+                                associated_source_id: _,
+                            }) if *refresh_table_id == refresh_args.table_id => {
+                                refresh_args.is_refreshing = true;
+                                tracing::info!(table_id = %refresh_table_id, "RefreshStart barrier received");
+                            }
```

Comment:
The is_refreshing being in memory should be ok, right? Since the upstream would try to emit RefreshStart again if errors occur.